feature-gate-threshold-automation init (#72)
solana-foundation/solana-improvement-documents: proposals/0072-feature-gate-threshold-automation.md

feature account to `Feature111111111111111111111111111111111111`.

This step is unchanged from its original procedure.

### Step 2: Staging Features for Activation

A multi-signature authority, comprised of key-holders from Solana Labs and
possibly other validator client teams in the future, will have the authority to
stage created features for activation.
In the future, this authority could be replaced by validator governance.

The multi-signature authority stages a feature for activation by invoking the
Feature Gate program's `StageFeatureForActivation` instruction. This instruction
expects:

- Data: The feature ID
- Accounts:
  - Staged Features PDA: writable
  - Multi-signature authority: signer 

One single PDA will be used to store two lists of features. Its data will be
structured as follows:

- 8 bytes for the current epoch.
- A fixed-size list of size 8 containing 32-byte feature IDs staged for
  activation at the end of the *current epoch*.
- A fixed-size list of size 8 containing the `u64` value of total stake support
  for each feature ID in the prior list.
- 8 bytes for the next epoch.
- A fixed-size list of size 8 containing 32-byte feature IDs staged for

>   activation at the end of the *next epoch*.

Reviewer: Could you specify this data structure in a C or Rust struct form? So that we know exactly how the data will be laid out.

Could we also specify the endianness of each field/list element?

Thanks so much!

Author: Yep, will do!

`StageFeatureForActivation` will add the provided feature ID to the **next
epoch's** set of staged features.

An example of how this account's state might work to coordinate both lists can
be found here:

<https://github.com/buffalojoec/feature-gate-data-poc/blob/master/src/lib.rs>

The Staged Features PDA could be derived simply from one literal seed:

> `"staged_features"`.

Reviewer: Instead of saying "could" can you say "will"? It's better to have a concrete proposal as a basis for discussion.

### Step 3: Signaling Support for Staged Features

With an on-chain reference point to determine the features staged for activation
for a particular epoch, nodes can signal their support for the staged features
supported by their software.

A node signals its support for staged features by invoking the Feature Gate
program's `SignalSupportForStagedFeatures` instruction. This instruction
expects:

> - Data: A `u8` bit mask of the staged features.

Reviewer: For completeness, do you mind explaining the reasoning for using a bit mask here? Is it to save on block space/network bandwidth?

Author: Bytes over the wire and program compute, really. Removes the need to send a list of public keys and then task the program with searching the list. The bitmask is already in order.

I can add this explanation.

Reviewer: Awesome thank-you!

- Accounts:
  - Staged Features PDA: writable
  - Vote account: signer

A `1` bit represents support for a feature. For example, for staged features
`[A, B, C, D, E, F, G, H]`, if a node wishes to signal support for all features
except `E` and `H`, their `u8` value would be 246, or `11110110`.

A node's submitted bit mask is then used to add that node's current epoch stake
to the Staged Features PDA for each feature for which it has signaled support.
This is done by using the Get Epoch Stake Syscall.

Nodes should send a transaction containing this instruction at some arbitrary
point during the epoch at least 128 slots before the end of the epoch and on

> startup after any reboot.

Reviewer: What happens if the transaction gets sent less than 128 slots before the end of the epoch? Does it get rejected?

Author: Yes. I'll add this detail.

> startup after any reboot.

Reviewer: If a node does not send the transaction at all, is that tallied the same as if the node had sent a transaction indicating it does not support any of the features?

Author: Yes. If their transaction is either rejected (too late in the Epoch) or never sent, their stake support is never tallied. If they signal "no support" for all features, their stake is also not tallied.

Should this be included in the proposal?

Reviewer: I think adding that reply to the SIMD would be great, just for completeness.

Note: If a feature is revoked, the list of staged features will not change, and
nodes may still signal support for this feature. However, the runtime will not
activate this feature if its corresponding feature account no longer exists
on-chain.

### Step 4: Feature Activation

During the epoch rollover, the runtime can simply load the Staged Features PDA
and calculate the stake - as a percentage of the total epoch stake - in support

> for each feature ID to determine which staged features to activate.

Reviewer: Does the runtime also need to check which features are revoked, and skip activating these? How would it do that (where would it look)?

Author: They wouldn't exist on-chain, so they're not in the Bank's feature set! Therefore something like `new_feature_activations.contains(..)` would resolve to `false`. This being an Agave-specific implementation bit, of course.

If you think we need to be more specific on this for specification purposes, I'm happy to add more context here.

Author: I do mention this earlier in the document, here:

https://github.com/solana-foundation/solana-improvement-documents/pull/72/files#diff-b798a8c9dca99c07b780074b5abdc2c8d81722af0a57dcdde447fbe96bbc8ad1R143-R146

But if you think this should move somewhere else I'm okay with that.

Reviewer: I think re-iterating that the runtime will not activate revoked features in the Step 4 section would be great, thanks!

Only features whose stake support meets the required threshold are activated.

> This threshold shall be set to 95% initially, but future iterations on the

Reviewer: How is this threshold set? Is it hard-coded into the runtime?

Author: Correct!

Reviewer: Yep agree. Do you mind specifying this in the SIMD? Thanks so much

   software.
4. **Activation and Garbage Collection:** At the end of epoch N+1, the runtime
   activates the feature-gates that have the necessary stake support. At the
   same time, the runtime also archives activated feature accounts and PDAs no
   longer required by the process.

### Step 1: Feature Creation

The first step is creation of a feature account, done by submitting a
transaction containing System instructions to fund, allocate, and assign the
feature account to `Feature111111111111111111111111111111111111`.

### Step 2: Staging Features for Activation

A multi-signature authority, comprised of key-holders from Solana Labs and
possibly other validator client teams in the future, will have the authority to
stage created features for activation.
In the future, this authority could be replaced by validator governance.

The multi-signature authority stages a feature for activation by invoking the
Feature Gate program's `StageFeatureForActivation` instruction. This instruction
expects:

- Data: The feature ID
- Accounts:
  - Staged Features PDA: writable
  - Multi-signature authority: signer 

`StageFeatureForActivation` will add the provided feature ID to the **next

> epoch's** Staged Features PDA.

Reviewer: What if we need to activate a feature urgently, for a security fix?

Reviewer: Not sure if this is out of scope or not, but one thought I had was that nodes could regularly check to see if their supported features are different to what their Support Signal PDA indicates, and send a transaction if this is the case.

Author: In order to add a new feature gate, the software would need to be updated, so the nodes would restart and then send another transaction with their support signal, so that new feature would be covered.

As far as bypassing the required threshold, I think we'd need a way to set the minimum required threshold per-feature, with the default being 95%. I had this as a side note for a future proposal, but if it's relevant enough I think it can be added to this one.

`[A, B, C, D, E, F, G, H]`, if a node wishes to signal support for all features
except `E` and `H`, their `u8` value would be 246, or `11110110`.

A node's submitted bit mask is then stored in a Support Signal PDA derived from
that node's vote address. The proposed seeds are defined below.

```
"support_signal" + <vote_address>
```

Nodes should send a transaction containing this instruction at some arbitrary
point during the epoch at least 128 slots before the end of the epoch and on
startup after any reboot.

Note: If a feature is revoked, the list of staged features will not change, and
nodes may still signal support for this feature. However, the runtime will not
activate this feature if its corresponding feature account no longer exists
on-chain.

### Step 4: Activation and Garbage Collection

During the epoch rollover, the runtime uses the validator support signals to
determine which staged features to activate.

To do this, the runtime walks all of the vote accounts, derives their Support
Signal PDA to read their bit mask, and tallies up the total stake support for
each staged feature. The runtime will also zero-out each bit mask, resetting

> each Support Signal PDA for the next epoch.

Reviewer: I think it would be safer to put as much of the logic on-chain as possible, to minimise the chance of a partition.

Instead of the runtime walking the vote accounts, the tally could be updated on-chain incrementally, in an account owned by the feature program. Then at the start of each epoch, the runtime can just look up the thresholds from this account.

Author: This is interesting, except the only issue is that the Feature Gate program has no way to know who _didn't_ vote, so there might still be a walking step in the runtime anyway.

Reviewer: If we expose `StakeHistory` to BPF programs, which we will likely have to do as part of the migration, the Feature Gate program can use that to determine which proportion of the stake has voted.

Author: Alternatively, if we can't expose it, we can see about this approach:
https://github.com/solana-labs/solana-program-library/pull/6106

Author: Or https://github.com/solana-labs/solana/pull/35252

Reviewer 2: @topointon-jump , how are you envisioning the program would calculate the specific stake amount/percentage of a vote account? The `StakeHistory` sysvar only contains the stake totals for the epoch (effective, activating, deactivating), and a vote account doesn't know how much stake is delegated to it.
I haven't thought of a way to move that tally on-chain without a massive undertaking.

Reviewer: @CriesofCarrots just for my understanding, why do we need the specific stake amount/percentage of a vote account? I thought all we needed to do was work out what proportion of the current stake for an epoch had sent support signals. We can use StakeHistory to tell us how much stake is activated for the epoch, and the Feature Gate program can update the on-chain tally as support signals are received, looking up each vote accounts delegated stake from the account values (not sysvars) as each support signal is received. 

Let me know if I'm not understanding something, also happy to hop on a call anytime :)

Reviewer 2: > looking up each vote accounts delegated stake from the account values

As I said, a vote account doesn't know how much stake is delegated to it, so the only way currently to do what you describe would be to pass every delegating stake account into the support transaction. That's obviously not feasible 😅 

Adding a new syscall that does `get_epoch_stake(validator_id: Pubkey) -> u64` as per https://discord.com/channels/428295358100013066/1141064343027462214/1214301700924510238 would enable tallying on a support transaction, however. Just trying to be clear that there is no existing mechanism to do this.

Author: > Adding a new syscall that does get_epoch_stake(validator_id: Pubkey) -> u64 as per ... would enable tallying on a support transaction, however.

@CriesofCarrots it sounds like this might be a good route to take in order to reduce complexity on the runtime.

How concerned are you with the O(n) tallying happening on epoch rollover? It seems like if creating this `get_epoch_stake` function isn't too much of a lift, it's definitely a more optimized solution.

Alternatively, there's also the block header route discussed in your linked Discord conversation.

Reviewer: Sorry @CriesofCarrots I misunderstood what you meant! Yes, imo adding the `get_epoch_stake` syscall is the easiest way to achieve this, if we go down this route.

Reviewer 2: @ripatel-fd , yes, I think that would be a great next step. Thank you!

Author: @ripatel-fd https://github.com/solana-foundation/solana-improvement-documents/pull/133

Author: The SIMD for the `getEpochStake` syscall was merged. Perhaps soon we can have another review on this?

Only features whose stake support meets the required threshold are activated.
This threshold shall be set to 95% initially, but future iterations on the
process could allow feature key-holders to set a custom threshold per-feature.

If a feature is not activated, either because it has been revoked or it did not
meet the required stake support, it must be resubmitted according to Step 2.

To ensure this new process doesn't overload the Feature Gate program's owned
accounts, during the activation stage, garbage collection will:

- Archive any activated feature accounts
- Archive this epoch's Staged Features PDA

> The runtime "archives" an account by assigning it to the Feature Tombstone.

Reviewer: Could the feature program manage this archiving process, instead of the runtime? Trying to avoid adding complexity to the runtime.

Author: Technically yes, but we'd need to invoke it somehow. Perhaps the first leader of the next epoch? I'm not sure on the complexity tradeoff here though.

Reviewer: The Support Signals can "crank" this. Over time the (small) cost would be spread out over all validators.

Author: I think this would be hard to line up. I'm a bit more inclined to leave the cleanup to the runtime, post-activation(s), and then explore tallying stake within the program as you mentioned. I think that's a good reduction in complexity. What do you think?

Author: As discussed in https://discord.com/channels/428295358100013066/1141064343027462214/1214300436614484149, I'm going to take a look at revising this proposal to reuse existing PDAs. This should address your concerns on garbage collection @topointon-jump.

Including the epoch in the account state was [previously mentioned here](https://github.com/solana-foundation/solana-improvement-documents/pull/72#discussion_r1364769365), and might be a good way to eliminate garbage collection. Since we've aligned on a fixed `u8` bitmask, this would allow us to still have a static-sized account.

Author: So, something like this would work to eliminate the need to create a new Staged Features PDA every epoch.
https://github.com/buffalojoec/feature-gate-data-poc/blob/master/src/lib.rs

However, I can't think of a feasible way to garbage collect feature accounts whose features have been activated on all clusters.

I think if we were to go the `get_epoch_stake(validator_id: Pubkey) -> u64` route as [discussed above](https://github.com/solana-foundation/solana-improvement-documents/pull/72#discussion_r1511779881), we wouldn't need the runtime to walk any accounts under `Feature11111..`.

Not having to walk the accounts means we can worry less about garbage collection. The approach I've shared for the Staged Features PDA can therefore be completely up to our discretion as an implementation detail. I'd also like to point out that if we aren't walking accounts in the runtime, then we don't have to worry about anyone spamming the mechanism by assigning some massive number of accounts to `Feature11111..`. Another reason the syscall is probably a better solution.

Author: This has been added to the proposal, in conjunction with the new syscall proposed in the other discussion thread.

---
simd: '0072'
title: Feature Gate Threshold Automation
authors:

>   - Tyera Eulburg

Author: My bad, can't believe you only noticed now!

2. The runtime automatically activates the feature on the next epoch boundary

The key-holder is the one who *manually* (with the help of the solana-cli)
assesses the amount of stake that recognizes the feature and decides whether
it is safe to activate. This is obviously brittle and subject to human error.

If instead the runtime was to assess this stake support for activating a
feature, this would eliminate the key-holder's responsibility to asses stake
support, reducing the risk of human error.

In a world where multiple clients will aim to push features and seek to agree on
their activation, a more automated and secure process will be extremely
beneficial.

## New Terminology

- **Feature Gate program:** The Core BPF program introduced in
  [SIMD 0089](https://github.com/solana-foundation/solana-improvement-documents/pull/89)
  that will own all feature accounts.
- **Staged Features PDA:** The PDA under the Feature Gate program used to track
  features submitted for activation.
- **Get Epoch Stake Syscall:** The new syscall introduced in
  [SIMD 0133](https://github.com/solana-foundation/solana-improvement-documents/pull/133)

>   that returns the current epoch stake for a given vote account address.

Reviewer 2: Update these links from the PRs to the actual SIMDs

   1, to be activated at the end of the *next epoch*.
3. **Signaling Support for Staged Features:** During the next epoch (epoch N+1),
   validators signal which of the staged feature-gates they support in their
   software.
4. **Feature Activation:** At the end of epoch N+1, the runtime activates the
   feature-gates that have the required stake support.

### Step 1: Feature Creation

The first step is creation of a feature account, done by submitting a
transaction containing System instructions to fund, allocate, and assign the
feature account to `Feature111111111111111111111111111111111111`.

This step is unchanged from its original procedure.

### Step 2: Staging Features for Activation

A multi-signature authority, comprised of key-holders from Solana Labs and
possibly other validator client teams in the future, will have the authority to
stage created features for activation.
In the future, this authority could be replaced by validator governance.

The multi-signature authority stages a feature for activation by invoking the
Feature Gate program's `StageFeatureForActivation` instruction. This instruction

> expects:

Reviewer 2: This SIMD refers to a couple Feature Gate program instructions: `StageFeatureForActivation` and `SignalSupportForStagedFeatures`. However, these instructions do not exist in the BPF program that SIMD-0089 proposed.
I think the language in this SIMD needs to make it clear that these are new proposed instructions for the Feature Gate BPF program.

One single PDA will be used to store two lists of features. Its data will be
structured as follows:

```c
#define FEATURE_ID_SIZE 32
#define MAX_FEATURES 8

/**
 * A Feature ID and its corresponding stake support, as signalled by validators.
 */
typedef struct {
    /** Feature identifier (32 bytes). */
    uint8_t feature_id[FEATURE_ID_SIZE];
    /** Stake support (little-endian u64). */
    uint8_t stake[8];
} FeatureStake;

/**
 * Staged features for activation. 
 */
typedef struct {
    /** The current epoch (little-endian u64). */
    uint8_t current_epoch[8];
    /**
     * Features staged for activation at the end of the current epoch, with
     * their corresponding signalled stake support.
     */
    FeatureStake current_feature_stakes[MAX_FEATURES];

    /** The next epoch (little-endian u64). */
    uint8_t next_epoch[8];
    /** Feature IDs staged for the _next_ epoch. */
    uint8_t next_features[MAX_FEATURES][FEATURE_ID_SIZE];
} StagedFeatures;
```

`StageFeatureForActivation` will add the provided feature ID to the **next
epoch's** set of staged features.

An example of how this account's state might work to coordinate both lists can
be found here:

> <https://github.com/buffalojoec/feature-gate-data-poc/blob/master/src/lib.rs>

Reviewer 2: Remove, especially since it looks like the POC is out of date. If there is anything important in the POC implementation, please call it out in prose.

program's `SignalSupportForStagedFeatures` instruction. This instruction
expects:

- Data: A `u8` bit mask of the staged features.
- Accounts:
  - Staged Features PDA: writable
  - Vote account: signer

A bit mask is used as a compressed ordered list of indices. This has two main
benefits:

- Minimizes transaction size for nodes, requiring only one byte to describe
  256 bytes (8 * 32) of data. The alternative would be sending `n` number of
  32-byte addresses in each instruction.
- Reduce compute required to search the list of staged features for a matching
  address. The bitmask's format provides the Feature Gate program with the
  indices it needs to store supported stake without searching through the staged
  features for a match.

A `1` bit represents support for a feature. For example, for staged features
`[A, B, C, D, E, F, G, H]`, if a node wishes to signal support for all features
except `E` and `H`, their `u8` value would be 246, or `11110110`.

A node's submitted bit mask is then used to add that node's current epoch stake
to the Staged Features PDA for each feature for which it has signaled support.

> This is done by using the Get Epoch Stake Syscall.

Reviewer 2: I think it would be helpful to make this paragraph more explicit. Something like: the `SignalSupportForStagedFeatures` processor passes the provided vote account address to the Get Epoch Stake syscall to determine that node's current epoch stake in lamports. The processor then adds this value to the `FeatureStake::stake` value for each feature marked as `1` (supported).

...but worded better.

Nodes are required to submit a transaction containing this instruction:

- Any time at least 4500 slots before the end of the epoch.
- During startup after any reboot.

Transactions sent too late (< 4500 slots before the epoch end) will be rejected
by the Feature Gate program.

If a node does not send this transaction or it is rejected, their stake is not

> tallied. This is analogous to a node sending this transaction in a valid point

Reviewer 2: ```suggestion
tallied. This is analogous to a node sending this transaction at a valid slot
```

in the epoch signalling support for zero features.

If a feature is revoked, the list of staged features will not change, and nodes
may still signal support for this feature. However, the runtime will not
activate this feature if its corresponding feature account no longer exists
on-chain.

### Step 4: Feature Activation

> During the epoch rollover, the runtime can simply load the Staged Features PDA

Reviewer 2: ```suggestion
During the epoch rollover, the runtime must load the Staged Features PDA
```

and calculate the stake - as a percentage of the total epoch stake - in support
for each feature ID to determine which staged features to activate.

> Only features whose stake support meets the required threshold are activated.

Reviewer 2: ```suggestion
Every feature whose stake support meets the required threshold must be activated.
```

This threshold will be hard-coded in the runtime to 95% initially, but future
iterations on the process could allow feature key-holders to set a custom
threshold per-feature.

As mentioned previously, if a feature was revoked, it will no longer exist
on-chain, and therefore will be not activated by the runtime, regardless of
calculated stake support.

If a feature is not activated, either because it has been revoked or it did not
meet the required stake support, it must be resubmitted according to Step 2.

Once the epoch rollover is complete, as soon as the Feature Gate program is
invoked again - either via `StageFeatureForActivation` or
`SignalSupportForStagedFeature` - the Staged Features PDA's account state is

> updated and the process begins again.

Reviewer 2: Please detail what it means for the Staged Features PDA's account state to be
updated.

The key-holder is the one who (with the help of the solana-cli) assesses the
amount of stake that recognizes the feature and decides whether it is safe to
activate. This is obviously brittle and subject to human error.

This SIMD proposes that the runtime itself replaces the human to assess the
amount of stake that supports a particular queued feature and only
automatically activates it when a preset threshold is met. 

## New Terminology

- Feature Gate program: The BPF program that will own all feature accounts.
- Feature-Request account: The PDA under the Feature Gate program used to track
features requested for activation.
- Feature-Queue account: The PDA under the Feature Gate program used to track
features queued for activation that must have support signaled by nodes.

## Detailed Design

Before this SIMD is published, we will deploy a BPF program to manage feature
activations, which will allow us to revoke pending feature activations
(previously not possible).
The feature activation process will otherwise be unchanged, so no SIMD is
required.
This program will be upgradeable; the authority and process through which it

> will be upgraded is outside the scope of this SIMD. Since this SIMD requires an

Reviewer 3: Agree with @mvines, we need another SIMD to define this program and interactions with it. In particular it needs to layout:
- What are the instructions to this new program, their inputs, and the side effects. This should probably look like [the system program docs](https://docs.rs/solana-program/1.17.1/solana_program/system_instruction/enum.SystemInstruction.html).
- Who can create new features? I can imagine that anyone can create a new feature account but the clients must recognize it. Maybe clients should vote on if a feature is valid? Maybe stake should vote on the intent to create new features? Lots to consider there. Also consider that allowing anyone to create a feature allows them to potentially slow down the runtime as we would need to process all of these feature accounts (what would you do if there were 500k feature accounts!).
- How are features which are "fully ratified" (i.e. everyone has had them active for a while) reaped? We want to eventually be able to remove feature gates from the code as they can have performance and readability impacts. How will we be able to safely remove them and what are the criteria for removal?
- How are features initialized at genesis? This is really important for testing!

Author: Done! #88 #89

update to the program, the upgrade details will be determined in parallel.

The initial Feature Gate Program can be found in this pull request to the
Solana Program Library:
<https://github.com/solana-labs/solana-program-library/pull/4987>.
The runtime changes and feature activation that will enable this program can be
found in this feature tracking issue:
<https://github.com/solana-labs/solana/issues/33547>.

### Overview

Consider two arbitrary versions of the Solana runtime, where the newer version
has more features in its feature set.

| v1.14.25  | v1.16.0   |
| --------- | --------- |
| Feature A | Feature A |
|           | Feature B |
| Feature C | Feature C |

> |           | Feature D |

Reviewer 3: What if there are a few "lone dissenters", should they be able to stall the whole network's progress?

> The runtime should activate each of these features when a strong majority of

Reviewer 3: This should probably be remove from the proposal if the goal is not to activate features.

Author: This could maybe re-worded instead. The idea here is it's going to leave off on the "OK this feature has enough support" and once that box is checked, hand things off to the process we are familiar with today after someone has manually queued a feature for activation.

See @CriesofCarrots comment [here](https://github.com/solana-foundation/solana-improvement-documents/pull/72#discussion_r1367146073)

the network supports it.
"Strong majority" in this context is defined by the current feature activation
process as 95% of stake.
“Support” for a feature means that a node is running a version of the Solana
runtime where the code required for the feature is present. For instance, on a
cluster with 50% of stake running v1.14.25 and 50% of stake running v1.16.0,
only Features A and C should be eligible for activation.

This SIMD proposes 3 steps over the span of one epoch to determine which
feature gates are queued for activation and assess stake support:

1. On an epoch boundary, the runtime generates a list of feature-gates queued
for activation.
2. Validators signal which of the queued feature-gates they support.

> 3. On the next epoch boundary, the runtime activates the feature-gates that

Reviewer 3: This should be removed from the proposal if the proposal's goal is detection of features

Reviewer 2: The goal of the proposal is to replace what is currently a human saying "yeah, I think enough of the validators recognize this feature that activation won't cause a partition in the network or piss anyone off too badly" with an automated process in the runtime. So yes, as with the current system, it ends with feature activation.

It is what you termed "Version Availability". If there is a substantive reason why this Version Availability check needs to happen before whatever governance system is adopted, instead of after as we've proposed, please explain.

PDA) will store the previous epoch’s requested activations.

When a new epoch begins, the Feature-Request PDA will be an empty list. When
key-holders submit the CLI command to activate a feature, the Feature Gate
Program appends the feature ID to the Feature-Request set at the same time as
creating the feature account . Revoked features are removed from the
Feature-Request list. At the end of the epoch, the Feature-Request set is
written to the feature queue PDA by the runtime, where it becomes immutable.
The Feature-Request PDA is then reset to an empty list again, also by the
runtime.

The newly created immutable Feature-Queue can then be used by nodes to signal
support for potential activation in the next epoch.

Note that for any given epoch, the epoch will begin with an immutable set of
previously requested feature activations, and end with the activation of those
with enough support. This process takes one epoch. Therefore, it will take at
least one more epoch than it currently does from the time a key-holder submits
a feature for activation via CLI to when it is actually activated.

### Signaling Support for Features

> 

Reviewer 3: Suppose I downgrade my node, mid-epoch for some reason (say, performance degradation). Can you downgrade your support?

Author: You'd be restarting after downgrading so you'd send a new signal transaction, with your updated bitmask!

We'd just want to make sure it always overwrites, which I can mention in the proposal.

With an on-chain reference point to determine the features queued for
activation, nodes can signal support for specific features in the queue.

> A node signals its support for a feature set by sending an instruction to the

Author: Could you be more specific?

As far as maximum length goes, it sounds like a static length per epoch is preferred. So `0`s all the way except for features you have.

> A node signals its support for a feature set by sending an instruction to the

Reviewer 3: so someone could spam that many features and then no more features are allowed?

> A node signals its support for a feature set by sending an instruction to the

Author: Ohh gotcha. Yeah we should design some kind of safeguard against this. Nice call!

> A node signals its support for a feature set by sending an instruction to the

Author: The multi-sig "queuer" introduced in the newest iteration should solve for this.

Feature Gate program, signed by their authorized voter, to store a record of
their supported feature set.
This supported feature set would be a bit mask of the list of queued features,
and it would be stored in a PDA derived from their vote address.
Nodes simply examine the Feature-Queue list, and mark a 1 for any features they
have in their feature set. The rest are marked 0.

In the previous example, a node running v1.14.25 would signal the following bit
mask:

```
1   0   1   0 
```

Similarly, a node running v1.16.0 would signal:

```

> 1   1   1   1 

Reviewer 3: They should just be allowed to send one some time in the epoch. The first slot can be very contentious.

```

> 

Reviewer 3: apologies this seems to be unaligned with actual section

Validators send these transactions signaling their support for a queued set of
features on the first slot of every epoch and on startup after any reboot.

Note: If a feature is revoked before it is moved to the immutable Feature-Queue
list, it is simply removed from the Feature-Request list and never presented to
nodes. However, if a feature is revoked after it becomes part of the immutable
Feature-Queue list, it remains in the list, and nodes still signal their
support if the feature ID is present in their software. However, the runtime
won’t activate this feature if its corresponding feature account no longer
exists on-chain.
This is how the runtime currently handles features and can be observed here:
<https://github.com/solana-labs/solana/blob/170478924705c9c62dbeb475c5425b68ba61b375/runtime/src/bank.rs#L8119-L8150>.

### Activating Supported Features

Once the epoch concludes, the runtime uses the validator support signals to

> activate the proper features.

Reviewer 3: Security fixes which introduce hard forks can be done out of band of this system

> activate the proper features.

Author: I think this is possibly worth exploring some more, even with the newest iteration of the proposal!

a guarantee that your feature will be activated when you run this command.

Instead, the runtime will decide through stake whether or not a feature should
be activated. This will be especially useful in a world where two separate
clients will aim to push features and want to agree on their activation.

## Alternatives Considered

### Representing the Current Feature Set (Alternative)

An alternate approach to the Feature Gate program’s on-chain storage is to
instead use one single PDA to store the queue and reset it every epoch. This
would allow us to use one less account and one less epoch, but it would make
things trickier when it comes to the timing of validators signaling support for
features.

For example, a key-holder might submit a feature for activation at a point late
in the epoch, so validators would need to be able to signal support for this
feature whenever it is submitted. This means validators would have to
consistently check the feature queue for changes, and possibly issue more
signals in the epoch.

This approach may increase the total required work for a node in a given epoch.

### Signaling Support for Features (Alternative)

> Rather than sending a transaction to the Feature Gate program, a node could

Author: I can add that sure, if it's something we want to consider.

> Rather than sending a transaction to the Feature Gate program, a node could

Reviewer 2: > What about a governance (i.e. manual, quorum based) voting alternative?

I think governance should come into play for *queuing* a feature for activation, ie a separate SIMD about how the Feature Gate Program works. The process outlined in this SIMD is more like a check to make sure each particular activation won't partition the network.

beneficial.

## New Terminology

- **Staging authority**: The authority who must sign to stage a given feature
  for activation, or unstage that feature.
- **Signal account (PDA)**: A PDA under the Feature Gate program used to
  prevent double-signaling.

## Detailed Design

A new feature activation process, comprised of the following steps:

1. Contributor creates a feature account and configures the feature's staging
   authority.
2. In some epoch `N-1`, the staging authority stages the feature for activation.
3. During the next epoch `N`, validators signal which of the staged features
   they support.
4. At the end of epoch `N`, the runtime activates the features with enough stake
   support.

### New Feature Account State

A new version of feature account state will track feature status, staging
authority, and stake support. Since the current feature state is a Rust
`Option<u64>`, we will use a `2` for the first byte to denote feature state v2.

```rust
enum FeatureV2Status {
    Inactive {
        staging_authority: Pubkey,
    },
    Staged { 
        staging_authority: Pubkey,
        stake_support: u64,

>     },

Reviewer 4: The staged state should also have the current epoch number so that at epoch boundaries we know that support was signaled in the previous epoch and not during an earlier epoch. This is important because the total stake can change epoch to epoch and affect whether at epoch boundaries a certain amount of stake is sufficient for activation.

>     Active,

Reviewer 4: We should also include which epoch or slot this feature was activated in

```

It contains just 8 bytes for the little-endian `u64` stake amount. These signal
accounts can serve as historical records for which nodes signaled support for
which features. They also serve as a source of truth for withdrawing support via
the `WithdrawSupportForStagedFeature` instruction, which will deduct the stored
stake value in the signal account from the feature's stake support and clear the
signal account.

```
Instruction: WithdrawSupportForStagedFeature
- Data:
    - 1 byte: Instruction discriminator
- Accounts:
    - [s]: Authorized voter
    - [ ]: Validator vote account
    - [w]: Signal account
    - [w]: Feature account
```

Although there is no requirement for features to be staged in the previous
epoch, staging might occur late in an epoch, after most support signals have
already been cast. This would result in not enough stake support until the next
epoch arrives and new signals are cast.

### Activation

At the end of the epoch, the runtime loads all staged feature accounts and
calculates their stake support as a percentage of the cluster-wide stake. Only
features whose account state has status `Staged` at the epoch rollover will be

> evaluated. 

Reviewer 4: I think this should specify that at the *beginning* of each epoch we need to load all staged feature accounts and calculate support. Typically we do epoch boundary operations at the beginning of a new epoch before processing transactions in the first block.

We should also specify where the cluster stake value comes from. Since the `GetEpochStake` syscall uses the "current epoch stakes", this should be the previous epoch's total stake. This value is calculated at the beginning of an epoch after activations and deactivations from the prior epoch have been applied, but before distributing inflation rewards.

Author: Sounds good I added this change!

Every feature whose stake support meets the required threshold will be
activated. This threshold will be hard-coded in the runtime to 95% initially,

> but future iterations on the process could make this threshold configurable.

Reviewer 4: I think this should also add details about how feature activation will be different with v2 feature gate accounts. Looks like currently we set the activated slot in v1 feature accounts when a feature gate is activated and store the updated account. But after this proposal we have a new `Active` enum variant to designate a v2 feature account as active.

We should also detail how the threshold calculation should be done so that it's consistent across all protocol implementations. Something like `100 * support_stake >= 95 * total_stake`?

Author: Added these details as well.

Features without the required stake support will remain `Staged` until they
acquire enough stake support or are manually unstaged.

## Alternatives Considered

## Impact

This new process for activating features directly impacts core contributors and
validators.

Core contributors will no longer bear the responsibility of ensuring the proper
stake supports their feature activation. However, this proposal does not include
a mechanism for overriding or customizing the stake requirement. This capability
should be proposed separately.

Validators will be responsible for signaling their vote using a transaction
which they've previously not included in their process. They also will have a
more significant impact on feature activations if they neglect to upgrade their

> software version.

Reviewer 4: nit: I don't think they have any more or less of an impact on feature activations after this proposal

on-chain, and therefore will be not activated by the runtime, regardless of
calculated stake support.

If a feature is not activated, either because it has been revoked or it did not
meet the required stake support, it must be resubmitted according to Step 2.

Once the epoch rollover is complete, the Feature Gate program will reset the
Staged Features PDA for the new epoch the very first time it is invoked
without error in the new epoch. Both instructions - `StageFeatureForActivation`
and `SignalSupportForStagedFeature` - will check the value for `next_epoch`
against the *current epoch* value in the `Clock` sysvar before executing the
steps outlined previously in this proposal. If a match is detected, the account
state is reset for the new epoch as follows:

1. The `current_feature_stakes` list is reset to an empty list.
2. The `current_epoch` is set to the value of `next_epoch`.
3. The feature IDs from the `next_features` list are written into the
   `current_feature_stakes` list with stake support initialized to zero.
4. The `next_epoch` value is set to the *next* upcoming epoch number
   (`current_epoch` + 1).

> 5. The `next_features` list is reset to an empty list.

Reviewer 2: Thanks, this is more clear. However, a question about the design: did you choose this "crank" approach so that the account update would be in the BPF program instead of a runtime responsibility?
I'm a little concerned about edge cases. It seems a lot easier to reason about current and next epoch (and related feature lists) when a reset occurs invariably on every epoch boundary. Otherwise, we foist additional logic on the validators.

For instance, validators need to query the `current_feature` list to determine their support data; but if the list may or may not have been updated yet, they have to look at both epochs and handle the logic correctly. In a similar vein, if there are no features staged for activation, there's no reason for validators to pay for a support-signal transaction at all.

What's the FD take on this @topointon-jump ?

Author: > did you choose this "crank" approach so that the account update would be in the BPF program instead of a runtime responsibility?

Yes, that's correct. @topointon-jump and I were leaning toward moving as much responsibility out of the runtime and into the program as possible. I was also aiming for zero garbage collection necessary.

I think a relationship where the runtime clears or otherwise updates account state, which the program has to then depend on, is brittle and should be avoided with these Core BPF programs wherever possible, especially across multiple clients.

I did entertain the idea of having one PDA that gets reused per-epoch to stage features, while using a PDA derived from each epoch to house the feature IDs validators must signal for, but I couldn't think of a good way to "crank" the transition from staged features to signal-ready features without the runtime's help.

Reviewer: @CriesofCarrots Yes exactly, the idea was to move as much complexity out of the runtime as possible so that it's easier to write new clients in the future.

Deciding whether or not to use a crank is a trade-off between the complexity introduced by having to deal with the un-cranked state, and the complexity introduced by pushing the account update and garbage collection logic inside the runtime.

We felt that using a crank would reduce the overall complexity on the runtime, but open to other thoughts 😄 do you think differently?

> In a similar vein, if there are no features staged for activation, there's no reason for validators to pay for a support-signal transaction at all.
I would say let's not worry about this case, it's far more likely for there to be staged features than not.

Reviewer 2: > We felt that using a crank would reduce the overall complexity on the runtime, but open to other thoughts 😄 do you think differently?

Yeah, I do think differently. I appreciate the desire to keep as much functionality in the program as possible, but the Feature Gate program is already a special case with regard to runtime responsibility. I think that having accounts the runtime depends on for activations possibly out of sync with the epoch rollover (because they haven't been cranked) is a far more brittle proposition than putting a small extra runtime requirement on clients.

> it's far more likely for there to be staged features than not.

I don't think this is true. Even during the craziest days of the Solana Labs client, the majority of epochs passed without a feature activation. Activations have slowed since then, due to greater care and scheduling around weekends. If anything, I would expect them to happen even less frequently as we move toward greater involvement by multiple parties and cluster governance.

How about we reconsider my original idea:

Use a new PDA for each epoch. The Feature Gate program can use the Clock sysvar to determine the current epoch and allow: (1) staging new features in the PDA for the current epoch; (2) signalling/tallying support for features in the PDA for the previous epoch.
Honestly, I don't think these PDAs ever need to be garbage collected; the number is very small, and they might be kind of nice for historical record purposes, especially if we allow variance of the minimum stake threshold for activation at some point in the future. But if we do feel strongly about them being garbage collected, we could add an unpermissioned instruction to close them that only executes successfully after the signalling/activation epoch has passed.

Seems like this satisfies your desire to keep logic in the program, it satisfies my desire to have runtime-dependency accounts in a known state, and it better satisfies my desire to not foist complex logic on the validators building signal transactions -- they would have to use the epoch number to derive the PDA address, but validators always know what epoch it is anyway.

I think this idea was set aside due to spam concerns, but seems like that isn't an issue with the existing design, since `StageFeatureForActivation` is gated.

Reviewer 2: @buffalojoec ^ Does this work for you?

Author: Yeah this works for me. As mentioned previously we no longer need to walk any accounts, so garbage collection is less of a concern.

> (1) staging new features in the PDA for the current epoch; (2) signalling/tallying support for features in the PDA for the previous epoch.

To be totally clear, this is one single PDA with two epochs of information (stage, signals), or two separate PDAs? Either is fine for me, since they're quite small, but two separate could be less brittle.

> But if we do feel strongly about them being garbage collected, we could add an unpermissioned instruction to close them that only executes successfully after the signalling/activation epoch has passed.

Yeah, nice idea. I'm fine with tabling this as a future revision when we may deem necessary, excluding it from this SIMD.

Reviewer 2: > To be totally clear, this is one single PDA with two epochs of information (stage, signals), or two separate PDAs? Either is fine for me, since they're quite small, but two separate could be less brittle.

I think we only need one PDA per epoch, with one epoch's-worth of information. (So I guess two, as you've phrased the question?) Let the epoch that's used in the PDA's address seed equal `N`. That PDA would be available to add (stage) new features during epoch `N`, and available to tally stake support during epoch `N+1`.

Or if we prefer the address seed to align with the signaling epoch (for validator signalling simplicity), we can stage in epoch `N-1` and signal in epoch `N`. I kind of like this better, now that I've typed it out.

Author: Yes I like the latter better as well. Stage in `N-1`, signal in `N`.

Author: I think the only weird part here is the creation of the PDA when it doesn't exist yet. The multisig "stager" would have to first fund this account if it doesn't exist. This can be implemented CLI/client-side, and the stager would pay the rent lamports on first feature staged, for every epoch.

Reviewer 2: I was envisioning that the optional PDA funding/creation would be part of the program, similar to how the ATA program works.

Reviewer 2: But yes, the first feature staged in an epoch would cost a little more. I think that's fine.

Author: Alrighty, let's see if I made it clear enough!

The first step is creation of a feature account, done by submitting a
transaction containing System instructions to fund, allocate, and assign the
feature account to `Feature111111111111111111111111111111111111`.

This step is unchanged from its original procedure.

### Step 2: Staging Features for Activation

A multi-signature authority, comprised of key-holders from Anza and
possibly other validator client teams in the future, will have the authority to
stage created features for activation.
In the future, this authority could be replaced by validator governance.

The multi-signature authority stages a feature for activation by invoking a new
Feature Gate program instruction: `StageFeatureForActivation`. This instruction
expects:

- Data: The feature ID
- Accounts:
  - Staged Features PDA: writable
  - Multi-signature authority: signer 

A PDA will be created for each epoch in which features are staged to be
activated. If no features are staged for a given epoch, that epoch's
corresponding PDA is not created.

These PDAs will not be garbage collected and can be referenced for historical
purposes.

When the first feature for an epoch is staged, the PDA is created. The

> `StageFeatureForActivation` processor will debit from the multisig enough

Reviewer 2: ```suggestion
`StageFeatureForActivation` processor will debit from a payer account enough
```

Author: Separate account?

Reviewer 2: Sure, that's fine. Just don't need to require it be the multisig.

Author: I can make this an optional signer on the instruction, and require it be provided only on the first staging.

  - Staged Features PDA: writable
  - Vote account: signer

A bit mask is used as a compressed ordered list of indices. This has two main
benefits:

- Minimizes transaction size for nodes, requiring only one byte to describe
  256 bytes (8 * 32) of data. The alternative would be sending `n` number of
  32-byte addresses in each instruction.
- Reduce compute required to search the list of staged features for a matching
  address. The bitmask's format provides the Feature Gate program with the
  indices it needs to store supported stake without searching through the staged
  features for a match.

A `1` bit represents support for a feature. For example, for staged features
`[A, B, C, D, E, F, G, H]`, if a node wishes to signal support for all features
except `E` and `H`, their `u8` value would be 246, or `11110110`.

The `SignalSupportForStagedFeatures` instruction processor will provide the
vote account's address to the `GetEpochStake` syscall to retrieve the stake
delegated to that vote account for the epoch. Then, using the `1` values
provided in the bitmask, the processor will add this stake value to each
corresponding feature ID's stake support in the Staged Features PDA.

> Just like the `StageFeatureForActivation` instruction, the Clock sysvar will be
> used to ensure the Staged Features PDA corresponding to the *current* epoch was
> provided.

Reviewer 2: ```suggestion
Similar to the `StageFeatureForActivation` instruction, the Clock sysvar will be
used to ensure the Staged Features PDA corresponding to the *current* epoch `N` was
provided.
```

- Initializes the `signal_epoch` to the current epoch + 1.
- Initializes the `stake_support` to zero.
- Copies the `staging_authority` from the feature's previous `Inactive` status.

```
Instruction: StageFeature
- Data:
    - 1 byte: Instruction discriminator (`2`)
- Accounts:
    - [s]: Staging authority
    - [w]: Feature account
```

This operation can be reversed anytime, as long as the feature still has
`Staged` status, via the `UnstageFeature` instruction.

```
Instruction: UnstageFeature
- Data:
    - 1 byte: Instruction discriminator (`3`)
- Accounts:
    - [s]: Staging authority
    - [w]: Feature account
```

### Signaling Support

Each epoch, all nodes must load all feature accounts with feature state v2 and
signal support for any staged features they wish to see activated. Staged
features they do not wish to activate can be intentionally omitted.

Support must be signaled through a single transaction. Depending on the number
of staged features in a given epoch, lookup tables can also be used. The
`SignalSupportForStagedFeatures` instruction requires nodes to include multiple
feature accounts after the first three required accounts.

```

> Instruction: SignalSupportForStagedFeatures

Author: Yeah, former. I had no intention of special-casing them.

    - 1 byte: Instruction discriminator (`4`)
- Accounts:
    - [s]: Authorized voter
    - [w]: Validator support signal account
    - [ ]: Validator vote account
    - ..n [w]: `n` Feature accounts
```

The authorized voter signer must match the authorized voter stored in the vote
account's state.

The program processor will use the `SolGetEpochStake` syscall to retrieve the
corresponding epoch stake for the provided vote account, and add it to the
feature's stake support. Whenever the Clock epoch has advanced beyond the
feature account's `signal_epoch`, that value is first updated to the Clock
epoch and stake support is reset to zero before tallying the current signal.

The signal account is a PDA derived from the validator vote account address
to prevent double-signaling.

```
"support_signal" + vote_address
```

```rust
/// Feature support signal state.
///
/// All `u64` integers are represented in little-endian.
struct SupportSignal {
    /// The epoch this node last submitted a support signal.
    epoch: u64,
}
```

The `epoch` value serves as a reference point for preventing double-signaling
by ensuring the `epoch` value is less than the current Clock epoch at the time

> of signaling.

Reviewer 5: You mean having a list of features that you are signaling for inside the `struct SupportSignal` right?

> of signaling.

Reviewer: @buffalojoec do you think it's likely that we start activating multiple features per epoch? I think it's unlikely that the feature cadence increases much more than it is today. Also, rent is being lowered so I don't think the fees should be that much of a concern. This could actually result in less work being done - a validator only has to send one support signal per feature, not per epoch.

    /** Stake support (u64 serialized to little-endian). */
    uint8_t stake_support[8];
} FeatureStake;

/**
 * Staged features for activation. 
 */
typedef struct {
    /**
     * Features staged for activation at the end of the current epoch, with
     * their corresponding signalled stake support.
     */
    FeatureStake features[MAX_FEATURES];
} StagedFeatures;
```

As depicted in the above layout, a maximum of 8 features can be staged for a
given epoch.

### Step 3: Signaling Support for Staged Features

With an on-chain reference point to determine the features staged for activation
for a particular epoch, nodes will signal their support for the staged features
supported by their software.

A node signals its support for staged features by invoking another new Feature
Gate program instruction: `SignalSupportForStagedFeatures`. This instruction
expects the Validator Support Signal PDA (defined below) to either exist or be
funded with enough rent-exempt lamports to initialize state. The processor will
allocate, assign, and initialize the account if it does not exist.

The `SignalSupportForStagedFeatures` instruction is structured as follows:

> - Data: A `u8` bit mask of the staged features.

Reviewer: Does this also need to include the epoch? As the list of staged features changes epoch-to-epoch, we need a way of ensuring that the bit mask in each validator support signal transaction is interpreted for the correct epoch. If the transaction is delayed or is executed in a different epoch than the one it was sent, for whatever reason, then the support signal will be mis-interpreted.

Reviewer 6: I would guess the `Staged Features PDA` supplied should indicate the intended epoch?

Author: It's in the PDA for the validator's support signal. When they send a bitmask, it gets recorded in their signal PDA, and if they send again, it's overwritten. The current epoch is checked via `Clock` sysvar when the program is invoked with this instruction.

  - Staged Features PDA: writable
  - Multi-signature authority: signer 

A PDA will be created for each epoch in which features are staged to be
activated. If no features are staged for a given epoch, that epoch's
corresponding PDA is not created.

These PDAs will not be garbage collected and can be referenced for historical
purposes.

When the first feature for an epoch is staged, the PDA is created. The
`StageFeatureForActivation` processor will debit from the multisig enough
lamports to allocate the new Staged Features PDA.

The address of the Staged Features PDA for a given epoch is derived as follows:

```

> "staged_features" + < epoch number >

Reviewer 2: Epoch number as little-endian u64?

Nodes should send a transaction containing this instruction at some arbitrary
point during the epoch at least 128 slots before the end of the epoch and on

> startup after any reboot. Transactions sent too late (< 128 slots before epoch

Reviewer: I think this < should be a > ?

Author: Nope I think it's correct!

Reviewer 7: Two things:
- Epochs can be shorter than 128 slots, I think the one after genesis starts with 32 slots or so.
- This is (128 * 0.4s = 51.2s) less than a minute. Our cache preparation phase is currently double that, and might expand further. I would make it a lot longer like 30 minutes (30*60s/0.4s = 4500 Slots).

Author: > * Epochs can be shorter than 128 slots, I think the one after genesis starts with 32 slots or so.

I think since this process is for new feature activations, it's safe to omit the first epoch after genesis in this design, right?

> * This is (128 * 0.4s = 51.2s) less than a minute. Our cache preparation phase is currently double that, and might expand further. I would make it a lot longer like 30 minutes (30*60s/0.4s = 4500 Slots).

Ok, I think a 4500 slot grace period is acceptable. I think that still leaves plenty of time for all nodes to submit their signals. It's worth mentioning that this would be a program design spec, which can be tweaked in the future.

to the Staged Features PDA for each feature for which it has signaled support.
This is done by using the Get Epoch Stake Syscall.

Nodes are required to submit a transaction containing this instruction:

- Any time at least 4500 slots before the end of the epoch.
- During startup after any reboot.

Transactions sent too late (< 4500 slots before the epoch end) will be rejected

> by the Feature Gate program.

Reviewer 2: Why do we need any deadline/restriction? I see the program cache being alluded to, but I don't see any specification that changes to the Staged Features PDA must begin or conclude at some point earlier than the epoch boundary. It's not obvious why the Feature Gate program can't continue to modify the Staged Features PDA up to the epoch boundary.

Author: Are you suggesting to explicitly explain the reasoning for this constraint in the SIMD, or are you asking why we actually need such a constraint? In either case, I agree it should be more explicit in the proposal.

It seems to me that @Lichtso intends to increase the cache preparation phase's initiation to be sooner in the epoch ([comment](https://github.com/solana-foundation/solana-improvement-documents/pull/72#discussion_r1603757414)), so we probably need a more concrete explanation as to why the Feature Gate program should have this constraint.

In this proposal, it was initially set to 128 slots remaining in the epoch because the cache preparation stage begins at a maximum of 128 slots before the end of the epoch, [as computed here](https://github.com/anza-xyz/agave/blob/9dd01f55182ff75a9c21bf89958ba5b37c8fe937/svm/src/transaction_processor.rs#L478-L482).

It's worth calling out that if we configure this constraint in the Feature Gate program and then later need to adjust it for an earlier cache preparation start, we'll be bound to using a feature gate for that change to the program.

Reviewer 2: I'm asking why we need a constraint at all. The cache being referenced is supposedly an agave-specific implementation, not part of the runtime protocol. If feature activation needs to work around any aspect of cache behavior, that seems like (a) a problem; and (b) a thing that definitely needs to be specified for all clients.

Author: There's really nothing stopping the cache from being able to anticipate the upcoming feature environment by querying the PDA under the Feature Gate program that would tell it what the current stake support is for each feature. It's actually arguably equally or more deterministic than the way it's done now.

The cache could run the calculation at it's current 128 slots before epoch boundary and probably get a pretty good understanding of which features have enough stake support. Increasing that slot buffer beyond 128 could just mean increasing the likelihood that not enough stake has been signaled for a feature. Perhaps this is a compromise the cache implementation should consider? After all, I believe recompilation is an optimization.

Author: I've added a commit removing the "slots remaining" constraint, since I think the above is a sensible direction to head in. If there's explicit opposition, it can be added back in.

Reviewer 7: @CriesofCarrots This is not only for the program cache and not only for our implementation. All validators must reach consensus before feature activation. If a feature is activated to close to the epoch boundary it is possible that multiple forks, some with the feature active others with it inactive, cross the boundary independently.

Reviewer 2: Features are always activated exactly on the epoch boundary, now and in this SIMD. I think maybe you mean something other than what you said?

Reviewer 7: Sorry for my imprecise formulation, I meant scheduled for activation.
If a feature is scheduled for activation in some slot close to the boundary, then it is possible that the TX is not yet rooted. Thus, some validator nodes will not have it scheduled, and will not activate it on the boundary.

Reviewer 2: Please explain how this is different from the cluster coming to consensus on other tx/block.

Reviewer 7: For features which only affect their own fork this is irrelevant. But for features which change validator global behavior this would break. And I am not sure if there is much to be gained from making exceptions or harving different feature types.

Author: Just adding an additional point of clarity here, if it helps.

Right now, in Agave, the cache is calling `compute_active_feature_set`, which will compute the feature set based on the on-chain accounts that have been created under `Feature11111...` (ie. "activated"). The cache does this ~128 slots before the epoch boundary to prepare the cache entries for the anticipated upcoming environment.

With the proposed system in place, the cache can still call `compute_active_feature_set` the same way it does now. The function will just compute the anticipated feature set from the proposed PDA (with stake support) rather than just the presence of on-chain accounts.

In almost every way it's the same as what we have now. The subtle difference is that someone can no longer activate a feature after the cache has done its feature-set computation, but nodes could still cast votes (signals) on a feature, bringing it to "eligible" status. This edge case exists today and just causes cache preparation/recompilation to be moot.

What @CriesofCarrots is suggesting is that we don't explicitly architect this mechanism around the cache, which is sensible, especially considering the cache preparation may change.

Author: @Lichtso do you still have concerns here? I'd like to move this SIMD along.
